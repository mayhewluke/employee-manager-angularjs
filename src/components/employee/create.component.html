<nav class="navbar navbar-default">
  <div class="container-fluid">
    <h2 class="pull-left">Add an employee</h2>
    <a class="btn btn-info navbar-btn pull-right" ui-sref="employees">Back</a>
  </div>
</nav>
<div class="container-fluid">
  <form class="form-horizontal" ng-submit="$ctrl.create($ctrl.employee)">
    <employee-form employee="$ctrl.employee" on-update="$ctrl.onUpdate"></employee-form>
    <div class="form-group">
      <div class="col-xs-10 col-xs-offset-2">
        <button type="submit" class="btn btn-success">Create</button>
      </div>
    </div>
    <div class="form-group">
      <p class="error alert alert-danger" ng-if="$ctrl.hasError">Something went wrong!</p>
    </div>
  </form>
</div>
