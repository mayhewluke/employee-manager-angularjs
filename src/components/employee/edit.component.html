<nav class="navbar navbar-default">
  <div class="container-fluid">
    <h2 class="pull-left">Edit employee</h2>
    <a class="btn btn-info navbar-btn pull-right" ui-sref="employees">Back</a>
  </div>
</nav>
<div class="container-fluid">
  <form class="form-horizontal">
    <employee-form employee="$ctrl.employee" on-update="$ctrl.onUpdate"></employee-form>
    <div class="form-group">
      <div class="col-xs-10 col-xs-offset-2">
        <button
          type="button"
          class="save btn btn-success"
          ng-click="$ctrl.save($ctrl.employeeUid, $ctrl.employee)"
        >
          Save
        </button>
        <button
          type="button"
          class="delete btn btn-danger"
          ng-click="$ctrl.remove($ctrl.employeeUid)"
        >
          Delete
        </button>
      </div>
    </div>
    <div class="form-group">
      <p class="error alert alert-danger" ng-if="$ctrl.hasError">Something went wrong!</p>
    </div>
  </form>
</div>
