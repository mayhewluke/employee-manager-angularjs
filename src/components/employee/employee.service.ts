import { default as uiRouterModule, StateService } from "@uirouter/angularjs";
import angular, { IPromise, IQService } from "angular";
import firebase from "firebase";

import { Employee } from "common/employeeTypes";

export class EmployeeService {
  // tslint:disable-next-line:no-shadowed-variable
  constructor(private $q: IQService, private $state: StateService) {}

  public fetchEmployees(): IPromise<{ [uid: string]: Employee }> {
    const { currentUser } = firebase.auth();
    if (currentUser === null) {
      this.$state.go("login");
      return this.$q.reject();
    }
    const ref = `/users/${currentUser.uid}/employees`;
    return this.$q
      .resolve(
        firebase
          .database()
          .ref(ref)
          .once("value"),
      )
      .then(snapshot => snapshot.val());
  }

  public fetch(uid: string): IPromise<Employee> {
    return this.fetchEmployees().then(
      employees => (employees[uid] ? employees[uid] : this.$q.reject()),
    );
  }

  public create(employee: Employee) {
    const { currentUser } = firebase.auth();
    if (currentUser === null) {
      this.$state.go("login");
      return this.$q.reject();
    }
    const ref = `/users/${currentUser.uid}/employees`;
    return this.$q((resolve, reject) =>
      firebase
        .database()
        .ref(ref)
        .push(employee)
        .then(x => resolve(x), e => reject(e)),
    );
  }

  public save(uid: string, employee: Employee) {
    const { currentUser } = firebase.auth();
    if (currentUser === null) {
      this.$state.go("login");
      return this.$q.reject();
    }
    const ref = `/users/${currentUser.uid}/employees/${uid}`;
    return this.$q.resolve(
      firebase
        .database()
        .ref(ref)
        .set(employee),
    );
  }
}

export const employeeServiceModule = angular
  .module("employee.service", [uiRouterModule])
  .service("EmployeeService", EmployeeService).name;
