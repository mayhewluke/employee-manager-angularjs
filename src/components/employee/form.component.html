<div class="form-group">
  <label for="name" class="col-xs-2">Name</label>
  <div class="col-xs-10">
    <input
      name="name"
      id="name"
      class="form-control"
      type="text"
      placeholder="Taylor"
      ng-model="$ctrl.employee.employeeName"
      ng-change="$ctrl.onUpdate($ctrl.employee)"
    />
  </div>
</div>
<div class="form-group">
  <label for="phone" class="col-xs-2">Phone</label>
  <div class="col-xs-10">
    <input name="phone"
      id="phone"
      class="form-control"
      type="text"
      placeholder="555-555-5555"
      ng-model="$ctrl.employee.phone"
      ng-change="$ctrl.onUpdate($ctrl.employee)"
    />
  </div>
</div>
<div class="form-group">
  <label for="shift" class="col-xs-2">Shift</label>
  <div class="col-xs-10">
    <select name="shift"
      id="shift"
      class="form-control"
      ng-model="$ctrl.employee.shift"
      ng-options="value as humanized for (value, humanized) in $ctrl.options"
      ng-change="$ctrl.onUpdate($ctrl.employee)"
    >
    </select>
  </div>
</div>
