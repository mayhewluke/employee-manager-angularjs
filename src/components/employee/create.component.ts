import {
  default as uiRouterModule,
  StateProvider,
  StateService,
} from "@uirouter/angularjs";
import angular, { IComponentOptions } from "angular";

import { Employee, ShiftDay } from "common/employeeTypes";

import { EmployeeService } from "./employee.service";

const config: IComponentOptions = {
  controller: class CreateEmployeeController {
    public employee: Employee = {
      employeeName: "",
      phone: "",
      shift: ShiftDay.Monday,
    };
    public hasError: boolean = false;

    constructor(
      // tslint:disable-next-line:no-shadowed-variable
      private EmployeeService: EmployeeService,
      private $state: StateService,
    ) {}

    public create(employee: Employee) {
      this.hasError = false;
      this.EmployeeService.create(employee)
        .then(() => this.$state.go("employees"))
        .catch(_ => (this.hasError = true));
    }

    public onUpdate = (employee: Employee) => {
      this.employee = angular.copy(employee);
    };
  },
  // TODO separate out template
  template: `
    <h2>Add an employee</h2>
    <form ng-submit="$ctrl.create($ctrl.employee)">
      <employee-form employee="$ctrl.employee" on-update="$ctrl.onUpdate"></employee-form>
      <button type="submit">Create</button>
      <p class="error" ng-if="$ctrl.hasError">Something went wrong!</p>
    </form>
  `,
};

const name = angular
  .module("createEmployeeComponent", [uiRouterModule])
  .component("createEmployee", config)
  .config(($stateProvider: StateProvider) => {
    $stateProvider.state("createEmployee", {
      component: "createEmployee",
      data: {
        requireAuth: true,
      },
      url: "/employee/create",
    });
  }).name;

export default name;
