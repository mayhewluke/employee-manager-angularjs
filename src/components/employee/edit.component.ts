import {
  default as uiRouterModule,
  StateProvider,
  StateService,
  Transition,
} from "@uirouter/angularjs";
import angular, { IComponentOptions } from "angular";

import { Employee, ShiftDay } from "common/employeeTypes";

import { EmployeeService } from "./employee.service";

const config: IComponentOptions = {
  bindings: { transition: "<" },
  controller: class EditEmployeeController {
    public employee: Employee = {
      employeeName: "",
      phone: "",
      shift: ShiftDay.Monday,
    };
    public employeeUid: string = "";
    public hasError: boolean = false;
    private transition!: Transition;

    constructor(
      // tslint:disable-next-line:no-shadowed-variable
      private EmployeeService: EmployeeService,
      private $state: StateService,
    ) {}

    public $onInit() {
      this.employeeUid = this.transition.params().employeeUid;
      this.EmployeeService.fetch(this.employeeUid).then(
        employee => (this.employee = employee),
      );
    }

    public save(uid: string, employee: Employee) {
      this.hasError = false;
      this.EmployeeService.save(uid, employee)
        .then(() => this.$state.go("employees"))
        .catch(_ => (this.hasError = true));
    }

    public remove(uid: string) {
      this.hasError = false;
      this.EmployeeService.remove(uid)
        .then(() => this.$state.go("employees"))
        .catch(_ => (this.hasError = true));
    }

    public onUpdate = (employee: Employee) => {
      this.employee = angular.copy(employee);
    };
  },
  // TODO separate out template
  template: `
    <h2>Edit {{ $ctrl.employee.employeeName }}</h2>
    <form>
      <employee-form employee="$ctrl.employee" on-update="$ctrl.onUpdate"></employee-form>
      <button type="button" class="save" ng-click="$ctrl.save($ctrl.employeeUid, $ctrl.employee)">Save</button>
      <button type="button" class="delete" ng-click="$ctrl.remove($ctrl.employeeUid)">Delete</button>
      <p class="error" ng-if="$ctrl.hasError">Something went wrong!</p>
    </form>
  `,
};

const name = angular
  .module("editEmployeeComponent", [uiRouterModule])
  .component("editEmployee", config)
  .config(($stateProvider: StateProvider) => {
    $stateProvider.state("editEmployee", {
      bindings: { transition: "$transition$" },
      component: "editEmployee",
      data: {
        requireAuth: true,
      },
      url: "/employee/:employeeUid",
    });
  }).name;

export default name;
