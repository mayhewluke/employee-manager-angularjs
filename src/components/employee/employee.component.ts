import {
  default as uiRouterModule,
  StateProvider,
  Transition,
} from "@uirouter/angularjs";
import angular, { IComponentOptions } from "angular";

import { EmployeeService } from "./employee.service";

const config: IComponentOptions = {
  bindings: { transition: "<" },
  controller: class EmployeeController {
    public employee = {};
    private transition!: Transition;

    constructor(
      private EmployeeService: EmployeeService, // tslint:disable-line:no-shadowed-variable
    ) {}

    public $onInit() {
      this.EmployeeService.fetch(this.transition.params().employeeUid).then(
        employee => (this.employee = employee),
      );
    }
  },
  template: `
  <p>Name: {{ $ctrl.employee.employeeName }}</p>
  <p>Phone: {{ $ctrl.employee.phone }}</p>
  <p>Shift: {{ $ctrl.employee.shift }}</p>
  `,
};

const name = angular
  .module("employeeComponent", [uiRouterModule])
  .component("employee", config)
  .config(($stateProvider: StateProvider) => {
    $stateProvider.state("employee", {
      bindings: { transition: "$transition$" },
      component: "employee",
      data: {
        requireAuth: true,
      },
      url: "/employee/:employeeUid",
    });
  }).name;

export default name;
