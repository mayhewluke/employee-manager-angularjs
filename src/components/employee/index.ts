import angular from "angular";

import createComponent from "./create.component";
import editComponent from "./edit.component";
import { employeeServiceModule } from "./employee.service";
import formComponent from "./form.component";
import listComponent from "./list.component";

export const employeeModule = angular.module("app.components.employee", [
  createComponent,
  editComponent,
  employeeServiceModule,
  formComponent,
  listComponent,
]).name;

export * from "./employee.service";
