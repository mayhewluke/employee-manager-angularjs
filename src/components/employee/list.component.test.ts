import angular, { ICompileService, IQService } from "angular";
import "angular-mocks";

import { Employee, ShiftDay } from "common/employeeTypes";
import { authModule } from "components/auth";
import { EmployeeService, employeeServiceModule } from "components/employee";

import listComponent from "./list.component";

describe("employee listComponent", () => {
  let render: (email?: string) => JQLite;
  let scope: any;
  let $q: IQService;
  let service: EmployeeService;
  const employees: { [uid: string]: Employee } = {
    uid1: { employeeName: "Taylor", phone: "555-5555", shift: ShiftDay.Monday },
    uid2: {
      employeeName: "Casey",
      phone: "123-456-7890",
      shift: ShiftDay.Friday,
    },
  };

  beforeEach(
    angular.mock.module(listComponent, authModule, employeeServiceModule),
  );
  beforeEach(
    angular.mock.inject(
      // tslint:disable:variable-name
      (
        $rootScope,
        $compile: ICompileService,
        _$q_: IQService,
        _EmployeeService_: EmployeeService,
        // tslint:enable:variable-name
      ) => {
        service = _EmployeeService_;
        $q = _$q_;
        scope = $rootScope.$new();
        render = () => {
          const ele = $compile(`<employees></employees>`)(scope);
          scope.$digest();
          return ele;
        };
        jest
          .spyOn(service, "fetchEmployees")
          .mockImplementation(() => $q(() => null));
      },
    ),
  );

  it(
    "requires authentication",
    angular.mock.inject(($state, $rootScope, AuthService) => {
      jest
        .spyOn(AuthService, "isAuthenticated")
        .mockImplementation(() => false);

      $state.go("employees");
      $rootScope.$apply();

      expect($state.current.name).not.toEqual("employees");
    }),
  );

  describe("when data has not yet loaded", () => {
    it("shows a loading message", () => {
      const element = render();

      expect(element.find(".loading")[0].outerHTML).toMatchSnapshot();
    });

    it("does not display the error message", () => {
      const element = render();

      expect(element.find(".error").length).toBe(0);
    });
  });

  describe("when employees are fetched", () => {
    beforeEach(() => {
      (service.fetchEmployees as any).mockImplementation(() =>
        $q.resolve(employees),
      );
    });

    it("lists the employees", () => {
      const element = render();

      expect(element.html()).toMatchSnapshot();
    });

    it("does not display the loading message", () => {
      const element = render();

      expect(element.find(".loading").length).toBe(0);
    });

    it("does not display the error message", () => {
      const element = render();

      expect(element.find(".error").length).toBe(0);
    });
  });

  describe("when fetching employees fails", () => {
    beforeEach(() => {
      (service.fetchEmployees as any).mockImplementation(() => $q.reject());
    });

    it("shows an error message", () => {
      const element = render();

      expect(element.find(".error")[0].outerHTML).toMatchSnapshot();
    });

    it("does not show the loading message", () => {
      const element = render();

      expect(element.find(".loading").length).toBe(0);
    });
  });
});
