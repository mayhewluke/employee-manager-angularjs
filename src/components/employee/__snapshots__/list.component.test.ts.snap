// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`employee listComponent when data has not yet loaded shows a loading message 1`] = `"<p class=\\"loading ng-scope\\" ng-if=\\"$ctrl.isLoading\\">Loading...</p>"`;

exports[`employee listComponent when employees are fetched lists the employees 1`] = `
"
    <h2>Employees</h2>
    <a ui-sref=\\"createEmployee\\">Add an employee</a>
    <!-- ngIf: $ctrl.isLoading -->
    <!-- ngIf: $ctrl.hasError -->
    <!-- ngRepeat: (uid, employee) in $ctrl.employees --><div ng-repeat=\\"(uid, employee) in $ctrl.employees\\" class=\\"ng-scope\\">
      <a ui-sref=\\"editEmployee({ employeeUid: uid })\\" class=\\"ng-binding\\">Taylor</a>
    </div><!-- end ngRepeat: (uid, employee) in $ctrl.employees --><div ng-repeat=\\"(uid, employee) in $ctrl.employees\\" class=\\"ng-scope\\">
      <a ui-sref=\\"editEmployee({ employeeUid: uid })\\" class=\\"ng-binding\\">Casey</a>
    </div><!-- end ngRepeat: (uid, employee) in $ctrl.employees -->
  "
`;

exports[`employee listComponent when fetching employees fails shows an error message 1`] = `"<p class=\\"error ng-scope\\" ng-if=\\"$ctrl.hasError\\">Something went wrong!</p>"`;
